colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,336 @@
+        return tensor.norm(2) / (tensor.numel() ** 0.5)
+
+    @staticmethod
+    def _approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col):
+        r_factor = (exp_avg_sq_row / exp_avg_sq_row.mean(dim=-1, keepdim=True)).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+
+    # approx_sq_grad for row parallel weight
+    @staticmethod
+    def _approx_sq_grad_row_parallel(exp_avg_sq_row, exp_avg_sq_col, sq_row_meam):
+        # row_meam = sq_row_meam
+        r_factor = (exp_avg_sq_row / sq_row_meam).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """
+        Performs a single optimization steps
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+        """
+        param_groups: Dict
+        {
+            "params":[weight, bias]
+            "lr"
+            "eps"
+            "clip_threshold"
+            "decay_rate"
+            "beta1"
+            "weight_decay"
+            "scale_parameter"
+            "relative_step"
+            "warmup_init"
+        }
+        """
+        for group in self.param_groups:
+            # update weight & bias
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = p.grad
+                if grad.is_sparse:
+                    raise RuntimeError("Adafactor does not support sparse gradients.")
+
+                state = self.state[p]
+                grad_shape = self.grad_shape_dict[id(p)]
+                param_is_dtensor = self.param_is_dtensor_dict[id(p)]
+                if param_is_dtensor:
+                    grad_shape = self.shard_to_param.get(id(p)).shape  # tp shape (2 dim)
+                factored, use_first_moment = self.factored_dict[id(p)], self.use_first_moment_dict[id(p)]
+                shard_spec = self.shard_spec_dict[id(p)]
+                if len(state) == 0:
+                    state["step"] = 0
+                    if use_first_moment:
+                        # Exponential moving average of gradient values
+                        state["exp_avg"] = torch.zeros_like(p)
+                    if factored and param_is_dtensor:
+                        if shard_spec.sharding_sequence[0] == "R":  # Col Parallel
+                            state["exp_avg_sq_row"] = torch.zeros(
+                                grad_shape[0] // self.data_parallel_size, device=p.device, dtype=p.dtype
+                            )  # [H/dp]
+                            state["exp_avg_sq_col"] = torch.zeros(
+                                grad_shape[1], device=p.device, dtype=p.dtype
+                            )  # [W/TP]
```

Comment:
Is this comment wrong? `grad_shape` is global shape (original shape), right?

Reply:
> Is this comment wrong? `grad_shape` is global shape (original shape), right?

Grad_shape is the param shape after tp;